src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1654,12 +1641,40 @@ public override BoundNode VisitCall(BoundCall node)
             }
 
             return builder == null ? default : builder.ToImmutableAndFree();
+
+            AttributeAnnotations removeInapplicableNotNullWhenSense(ParameterSymbol parameter, AttributeAnnotations annotations, bool sense)
+            {
+                var whenSense = sense ? AttributeAnnotations.NotNullWhenTrue : AttributeAnnotations.NotNullWhenFalse;
+                var whenNotSense = sense ? AttributeAnnotations.NotNullWhenFalse : AttributeAnnotations.NotNullWhenTrue;
+
+                // NotNullWhenSense (without NotNullWhenNotSense) must be applied on a bool-returning member
+                if ((annotations & whenSense) != 0 &&
+                    (annotations & whenNotSense) == 0 &&
+                    parameter.ContainingSymbol.GetTypeOrReturnType().SpecialType != SpecialType.System_Boolean)
+                {
+                    annotations &= ~whenSense;
+                }
+
+                // NotNullWhenSense must be applied to a reference or unconstrained generic type
+                if ((annotations & whenSense) != 0 && parameter.Type.IsValueType != false)
+                {
+                    annotations &= ~whenSense;
+                }
+
+                // NotNullWhenSense is inapplicable when argument corresponds to params parameter and we're in expanded form
+                if ((annotations & whenSense) != 0 && parameter.IsParams && expanded)
```

Comment:
>parameter.IsParams [](start = 54, length = 18)

This is still not accurate. Only the last parameter is subject to expansion. I recommend replacing this check with a check whether parameter is the last one. Please add a test exposing the flaw in the current implementation

Reply:
Thanks. 
That `IsParams` always trips me up. I expect it would only be true for proper `params`, rather than just parameters that have the `[ParamArray]` attribute.